### PR TITLE
Hardware keyboard improvements

### LINKED ISSUE #54: Caps lock doesn't work with TOHKBD
Turning on caps lock in TOHKBD doesn't affect FingerTerm. Normal shift works fine but caps lock just does nothing. Using latest 1.1.9 with TOHKBD package 0.5.4. No idea if this affects other external keyboards.

### PR DESCRIPTION
**Pull request by urjaman**
This fixes #54 along with other improvements to behavior of hardware keyboards:
- virtual/touch keyboard doesnt wake up on hardware keypress, for using hwkbd while vkb enabled 
- use event.text from Qt keyEvents, fixes capslock and russian (propably other too) text input
- modifiers (shift,alt,control) for editing/nav keys, notably alt-left/right
- insert key at all (also with modifiers)
- Function keys (with modifiers)

### CODE REVIEW

**Comment by kimmoli** [comment]
Tested, WFM

File: terminal.cpp [hunk 1]
```diff
@@ -174,19 +174,21 @@ void Terminal::putString(QString str, bool unEscape)
...
 }
 
-void Terminal::keyPress(int key, int modifiers)
+void Terminal::keyPress(int key, int modifiers, QString text)
 {
     QChar c(key);
 
     resetBackBufferScrollPos();
 
+    if ((!text.isEmpty()) && (text[0].unicode() == 0)) text.clear();
```
Comment (pvuorela) [inline terminal.cpp]: Looks strange case, not sure if Qt is really supposed to pass such string. KeyEvent documentation says text is empty with modifier keys. Anyway, maybe text[0].isNull() could be slightly clearer?

@giucam any chance QtWayland passes bad event data?

File: terminal.cpp [hunk 2]
```diff
@@ -213,18 +219,111 @@ void Terminal::keyPress(int key, int modifiers)
...
+        if( key==Qt::Key_Up )
+            toWrite += QString("%1%2A").arg(ch_ESC).arg(cursorModif).toLatin1();
+        if( key==Qt::Key_Down )
+            toWrite += QString("%1%2B").arg(ch_ESC).arg(cursorModif).toLatin1();
+        if( key==Qt::Key_Right )
+            toWrite += QString("%1%2C").arg(ch_ESC).arg(cursorModif).toLatin1();
+        if( key==Qt::Key_Left )
+            toWrite += QString("%1%2D").arg(ch_ESC).arg(cursorModif).toLatin1();
+        if( key==Qt::Key_PageUp )
+            toWrite += QString("%1[5~").arg(ch_ESC).toLatin1();
```
Comment (pvuorela) [inline terminal.cpp]: On earlier code above CSI was formed with ch_ESC and cursorModif. I have no idea why the name is cursorModif in the first place, but could expect for consistency control codes being formed the same way here. Better might be combining those two characters separately instead of individually for each case.

Comment (pvuorela) [inline terminal.cpp]: Oh right, old code being moved here. Well, anyway, these could have some cleanup :)

File: terminal.cpp [hunk 3]
```diff
@@ -213,18 +219,111 @@ void Terminal::keyPress(int key, int modifiers)
...
-        toWrite += QString("%1%2B").arg(ch_ESC).arg(cursorModif).toLatin1();
-    if( key==Qt::Key_Right )
-        toWrite += QString("%1%2C").arg(ch_ESC).arg(cursorModif).toLatin1();
-    if( key==Qt::Key_Left )
-        toWrite += QString("%1%2D").arg(ch_ESC).arg(cursorModif).toLatin1();
+    } else {
+        char modChar = '1' + modcode;
+
+        if( key==Qt::Key_Up )
+            toWrite += QString("%1[1;%2A").arg(ch_ESC).arg(modChar).toLatin1();
```
Comment (pvuorela) [inline terminal.cpp]: Not a fan of duplicating key handling with and without modifiers. How about something (not taking into account possible CSI comment earlier) like:

QString modifierString;
if (modCode > 0) {
   modifierString = QString(";") + modCode; 
}
[...]
toWrite += QString("%1%21%3A").arg(ch_ESC).arg(cursorModif).arg(modifierString).toLatin1();

Would though add explicit 1 for first parameter which is now missing in normal case.

File: terminal.cpp (hunk 1 above)
Comment (urjaman) [inline terminal.cpp]: I know only what i observe, text.count() ==1 and text[0].unicode() == 0 was the result for arrows/other non-letter keys... that broke my logic of assuming you can trust isEmpty()...

Maybe that could use a comment, i wrote all of this not expecting to try to upstream, started as a proof-of-concept bugfix idea.

File: terminal.cpp (hunk 3 above)
Comment (urjaman) [inline terminal.cpp]: You'd also need to turn off the effect of iAppCursorKeys (thats a terminal mode) for the case of the modifiers, and also
note not all of them are just normal with ;m (m=modifier) tacked in the middle, eg F1-F4, home/end (those were the first ones i noticed from the diff) are minus 'O' add '[1;m'  for normal vs with modifiers.

FYI: My quickly googled reference on the codes was https://code.google.com/p/mintty/wiki/Keycodes
So basically the splitting of modifiers/without was caused by following the table styles there (seperate codes for plain and modifiers) and i wanted to make sure i didnt break anything, but i know it could be more compact/smarter.
I'm not sure if it'd be actually of any real use to do that though.

Comment (pvuorela) [inline terminal.cpp]: > Explain? I dont understand what you meant by this bit.

https://en.wikipedia.org/wiki/ANSI_escape_code#CSI_codes
"The numbers are optional parameters. The default value used for omitted parameters varies with the command, but is usually 1 or 0"

So main case "esc [ A" should be equivalent with "esc [ 1 A", and I guess with even "esc [ 1; 1 A".

Not objecting separate branches anyway.

File: terminal.cpp (hunk 1 above)
Comment (pvuorela) [inline terminal.cpp]: Comment would be good.

File: terminal.h [hunk 4]
```diff
@@ -79,7 +79,7 @@ class Terminal : public QObject
 
     QList<TermChar>& currentLine();
 
-    Q_INVOKABLE void keyPress(int key, int modifiers);
+    Q_INVOKABLE void keyPress(int key, int modifiers, QString text);
```
Comment (pvuorela) [inline terminal.h]: Minor detail, but commonly const reference. With default value can avoid "" in virtual keyboard handling.

**Comment by urjaman** [comment]
Implemented some of the comments, made the null char hack unnecessary, reordered the parsing a bit...
The diff looks quite horrible, i suggest just looking at the resulting code :), and yeah i know that too could still use more work.

I'll rebase this and squash the unnecessary hacks etc away later if requested, just showing my incremental progress for comments (and without losing the old comments).

Note: somebody noted that ctrl-backspace doesnt work, i might implement that at some point and other modifiers for enter and backspace.

**Comment by urjaman** [comment]
ok i think thats it for now (until somebody comments something), ctrl-backspace and other similars implemented, cleaned up a bit.

**Comment by pvuorela** [comment]
LGTM. Could squash commits as you said, quite a lot of code churn on one method.

Russian keyboard handling still having small problems: e.g ctrl+z/c combinations don't do what they usually do while on desktop they work.

**Comment by urjaman** [comment]
Thanks, I'll squash them (atleast the fixup ones)  later when i get a moment.

Russian Ctrl-<key> handling: known, i dont really know how to do it in a beautiful way at the moment (without a LUT russian => qwerty-equivalent) since qt uses the unicode points as "key" instead of the underlying hw key - maybe look into the native scan code stuff...

**Comment by urjaman** [comment]
Done. It was all quite in the same area so after some consideration I squashed all the stuff on this one function to one commit, ended up better than I expected for such a big squash.
